android/modules/calendar/src/java/ti/modules/titanium/calendar/RecurrenceRuleProxy.java

```diff
@@ -0,0 +1,336 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2017 by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+package ti.modules.titanium.calendar;
+
+import org.appcelerator.kroll.KrollDict;
+import org.appcelerator.kroll.KrollProxy;
+import org.appcelerator.kroll.annotations.Kroll;
+import org.appcelerator.titanium.TiC;
+import org.appcelerator.titanium.util.TiConvert;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Kroll.proxy
+public class RecurrenceRuleProxy extends KrollProxy {
+
+	// Keys for daysOfTheWeek dictionary.
+	private final String dayOfWeekKey = "daysOfWeek";
+	private final String weekNumberKey = "week";
+
+	// Keys for end dictionary.
+	private final String until = "endDate";
+	private final String count = "occurrenceCount";
+
+	// Field with the Date of the event begin to provide parity with iOS.
+	// Some information exposed can't be get solely from the recurrence rule.
+	private Date eventBegin;
+	private String rRule;
+
+	//region values that can be set from a Creation Dictionary
+	private Integer frequency = null;
```

Comment:
Ideally, `frequency` should be an enum, not an `int` or `Integer`. This way we can easily constrain the value to known constants and provide easy mapping to know Android identifiers.

For example:

```java
public enum TiRecurrenceFrequencyType
{
	DAILY(CalendarModule.RECURRENCEFREQUENCY_DAILY, "DAILY"),
	WEEKLY(CalendarModule.RECURRENCEFREQUENCY_WEEKLY, "WEEKLY"),
	MONTHLY(CalendarModule.RECURRENCEFREQUENCY_MONTHLY, "MONTHLY"),
	YEARLY(CalendarModule.RECURRENCEFREQUENCY_YEARLY, "YEARLY");

	private final int tiIntId;
	private final String rfcStringId;

	private TiRecurrenceFrequencyType(int tiIntId, String rfcStringId)
	{
		this.tiIntId = tiIntd;
		this.rfcStringId = rfcStringId;
	}

	public int toTiIntId()
	{
		return this.tiIntId;
	}

	public String toRfcStringId()
	{
		return this.rfcStringId;
	}

	public static TiRecurrenceFrequencyType fromTiIntId(int value)
	{
		for (TiRecurrenceFrequencyType nextObject : TiRecurrenceFrequencyType.values()) {
			if ((nextObject != null) && (nextObject.tiIntId == value)) {
				return nextObject;
			}
		}
		return null;
	}

	public static TiRecurrenceFrequencyType fromRfcStringId(String value)
	{
		for (TiRecurrenceFrequencyType nextObject : TiRecurrenceFrequencyType.values()) {
			if ((nextObject != null) && (nextObject.rfcStringId == value)) {
				return nextObject;
			}
		}
		return null;
	}
}
```

With the above, you no longer need your `frequencyMap` static variable since the RFC frequency string ID mapping is already embedded into the enum. You can also easily convert/validate the JavaScript `frequency` field in your `RecurrenceRuleProxy` via the enum's `fromTiIntId()` like this...

```java
if (creationDictionary.containsKey(TiC.PROPERTY_FREQUENCY)) {
	int integerId = TiConvert.toInt(creationDictionary.get(TiC.PROPERTY_FREQUENCY));
	this.frequency = TiRecurrenceFrequencyType.fromTiIntId(integerId);
	if (this.frequency == null) {
		// Given frequency ID from JavaScript is invalid. Log an error.
	}
}
if (this.frequency == null) {
	// Default to daily if not assigned or given an invalid frequency.
	this.frequency = TiRecurrenceFrequencyType.DAILY;
}
```

You don't have to do this now, but Gary and I want to slowly transition our code to use the above pattern since it's a much more maintainable and less error prone solution. I've already started doing so with Titanium's orientation type handling here: [TiDeviceOrientation.java](https://github.com/appcelerator/titanium_mobile/blob/master/android/titanium/src/java/org/appcelerator/titanium/util/TiDeviceOrientation.java).

Reply:
@jquick-axway Updated.